### DISCUSSION TITLE
Run unit tests in interactive notebook?

### DISCUSSION THREAD

**Discussion by egil** [dotnet/interactive#1356] [Q&A]
Hi there,

Im the creator of bUnit, and would like to create a notebook that teaches how to use bUnit, and more generally, how to use xUnit/NUnit/MSTest.

What I cannot figure out is whether or not it is possible. This is what I imagine doing:

1. Have a code block with a system under test defined. This could either be in C# or razor code. For bUnit, that would be Blazor component defined in razor syntax that would need to be compiled.
2. Have another code block with the unit test inside it. It should reference previous code block that contains the system/component under test.
3. Be able to run the test(s) defined in the test code block, and have the console output from running the test, e.g. like `dotnet test` does.

There would be markdown in between the code blocks that explains what goes on in each block.

I this possible?

Thanks, Egil

**Comment by jonsequitur** [comment]
This should be possible. 

1. @plbonneville has an extension for Razor that might be useful here. 
2. This part should just work, assuming the user executes the cells in the right order. In C# scripting (used by .NET Interactive for C# support), each submission (i.e. cell execution) adds an in-memory assembly referencing the previous one, in a chain. So...
3. The test runner will have to be told how to scan these assemblies. They're not files on disk. You can access them at runtime, including via extension middleware or custom magic commands.

![image](https://user-images.githubusercontent.com/547415/119267516-d6a2f800-bba3-11eb-88d3-fc68ea7f0f0d.png)
 
A lot of the relevant pieces can be found here: https://github.com/dotnet/interactive/blob/4c03a63e813beb15866ef0a01b1712cc912eceb0/src/Microsoft.DotNet.Interactive.CSharp/InteractiveWorkspace.cs#L0-L1

**Comment by plbonneville** [comment]
Hi @egil,

Completely agree with your idea. Creating a Blazor extension was actually my first intent before creating my [Razor extension](https://github.com/plbonneville/RazorInteractive). The idea was to use a Jupyter notebook to teach Blazor.

I chose to focus on Razor realizing that a Razor extension had a raison d’être of its own.

I’d love to help you with this project.

Cheers,

**Reply by egil** [reply]
Hi @plbonneville

Awesome. I have some insights into the blazor renderer from my work with bUnit, but also think the folks building https://blazorrepl.com/ might know a thing or two about triggering the compiler.

Do you have any code you can share that shows how you do things with the Razor extension you are building?

**Reply by plbonneville** [reply]
Hi @egil,

I've added my source code to compile and render Razor components (.razor) in .NET Interactive Notebooks in its own repo [BlazorInteractive](https://github.com/plbonneville/BlazorInteractive). The project is available as a NuGet package.

I based the code on the [Blazor REPL](https://github.com/BlazorRepl/BlazorRepl). It had to replaced the BlazorRepl `CompilationService` class as theirs is designed to run as Blazor WebAssembly and fetch the references over the wire.

You'll see that I used bUnit to fetch the markup.

It works in VS Code with both `.dib` and `.ipynb` file format and in Jupyter.

**Reply by egil** [reply]
Ok, cool, I'll add it to my long list of things to look at.

**Comment by GeorgeS2019** [comment]
Definitely, a great way to go => Integrating Blazor visualization in .NET interactive

![msedge_lahSYjk5ug](https://user-images.githubusercontent.com/49812372/184486581-7bc073a8-a1b0-4a2a-a8bf-963cb381a926.gif)
